subsys/pm/pm.c

```diff
@@ -27,7 +27,24 @@ LOG_MODULE_REGISTER(pm, CONFIG_PM_LOG_LEVEL);
 	(COND_CODE_1(CONFIG_SMP, (arch_curr_cpu()->id), (_current_cpu->id)))
 
 static ATOMIC_DEFINE(z_post_ops_required, CONFIG_MP_MAX_NUM_CPUS);
-static sys_slist_t pm_notifiers = SYS_SLIST_STATIC_INIT(&pm_notifiers);
+
+/* Active state notifiers list */
+static sys_slist_t pm_notifiers_active = SYS_SLIST_STATIC_INIT(&pm_notifiers_active);
+
+/* Generate a notifier list for each DT power state */
+DT_FOREACH_CHILD_STATUS_OKAY_SEP(DT_PATH(cpus, power_states), PM_NOTIFIERS_CREATE_STATES, (;));
+
+/* Array of all notifier lists, active state as first element */
+static sys_slist_t *pm_notifiers_array[] = {
+	&pm_notifiers_active,
+	DT_FOREACH_CHILD_STATUS_OKAY_SEP(DT_PATH(cpus, power_states), PM_NOTIFIERS_STATES, (,))
+};
+
+/* Array of all DT power-states, active state as first element */
+static const struct pm_state_info pmstates[] = {
+	{ .state = PM_STATE_ACTIVE },
```

Comment:
question: why is active needed here?

Reply:
Active states are included in entry/exit notifications.

Here specifically, it mirrors the array contents and order of `pm_notifiers_array[]`.

Prepending the active state to both lets us use the index to refer to the same state between either, this is used to compare against the current system state and execute the correct notifier list, as well as during notifier un/registration.